kyo-core/shared/src/main/scala/kyo/tags.scala

```diff
@@ -1,41 +1,214 @@
+                    case ',' =>
+                        if opens == 0 then
+                            idx + 1
+                        else
+                            loop(opens, idx + 1)
+                    case ']' =>
+                        if opens == 0 then
+                            -1
+                        else
+                            loop(opens - 1, idx + 1)
+                    case '[' => loop(opens + 1, idx + 1)
+                    case _   => loop(opens, idx + 1)
+            loop(0, idx)
+        end nextParam
+
+        // Macro methods
+
+        def tagImpl[T: Type](using Quotes): Expr[Tag[T]] =
+            import quotes.reflect.*
+
+            val encoded = encodeType(TypeRepr.of[T])
+            '{ new Tag($encoded) }
+        end tagImpl
+
+        def encodeType(using Quotes)(tpe: quotes.reflect.TypeRepr): Expr[String] =
+            import quotes.reflect.*
+
+            tpe.dealias match
+                case AndType(_, _) | OrType(_, _) =>
+                    report.errorAndAbort(s"Unsupported Tag type ${tpe.show}")
+                case tpe if tpe.typeSymbol.fullName == "scala.Nothing" =>
+                    report.errorAndAbort(s"Unsupported Tag type ${tpe.show}")
+                case tpe if tpe.typeSymbol.isClassDef =>
+                    val sym  = tpe.typeSymbol
+                    val path = tpe.dealias.baseClasses.map(_.fullName).mkString(";") + ";"
+                    val variances = sym.typeMembers.flatMap { v =>
+                        if !v.isTypeParam then None
+                        else if v.flags.is(Flags.Contravariant) then Some("-")
+                        else if v.flags.is(Flags.Covariant) then Some("+")
+                        else Some("=")
+                    }
+                    val params = tpe.typeArgs.map(encodeType)
+                    if params.isEmpty then
+                        Expr(path)
+                    else
+                        concat(
+                            Expr(s"$path[") ::
+                                (variances.zip(params).map((v, p) => Expr(v) :: p :: Nil)
+                                    .reduce((a, b) =>
+                                        a ::: Expr(",") :: b
+                                    ) :+ Expr("]"))
+                        )
+                    end if
+                case _ =>
+                    tpe.asType match
+                        case '[tpe] =>
+                            Expr.summon[Tag[tpe]] match
+                                case None =>
+                                    report.errorAndAbort(
+                                        s"Please provide an implicit ${TypeRepr.of[Tag[tpe]].show}"
+                                    )
+                                case Some(value) =>
+                                    '{ $value.tpe }
+                        case _ =>
+                            report.errorAndAbort(s"Unsupported Tag type ${tpe.show}")
```

Comment:
Can we make the error messages unique to assist with debugging issues? When users submit tickets it may be useful to know which case they are hitting